### PR TITLE
OpenTap Login API

### PR DESCRIPTION
**Pull request by rmadsen-ks**
- added an API for managing access tokens
- Added cli actions that makes it possible to manage tokens from the command line.

### CODE REVIEW

**Comment by Denrask** [comment]
Hi @rmadsen-ks, can you make this build?

File: Engine/Login/LoginCliAction.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+                    var error = response.Content.ReadAsStringAsync().Result;
+                    throw new Exception("Unable to connect: " + response.StatusCode);
+                }
+
+                Site = Site ?? Url.Replace("keycloak.", "");
+                var responseString = response.Content.ReadAsStringAsync().Result;
+                var json = System.Text.Json.JsonDocument.Parse(responseString);
+                //"expires_in":300,"refresh_expires_in":1800
+                var accessExp = DateTime.Now.AddSeconds(300);
+                var refreshExp = DateTime.Now.AddSeconds(1600);
```
Comment (db-ks) [inline Engine/Login/LoginCliAction.cs:55]: 1600->1800?

File: Engine/Login/LoginCliAction.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        /// <summary> Executes the action. </summary>
+        public int Execute(CancellationToken cancellationToken)
+        {
+            var url = $"https://{Url}/auth/realms/{Realm}/protocol/openid-connect/token";
+            var http = new HttpClient { };
+            using (var content =
+                   new FormUrlEncodedContent(new Dictionary<string, string>
+                   {
+                       {"username", UserName},
+                       {"password", Password}, {"grant_type", "password"}, {"client_id", ClientId}
```
Comment (db-ks) [inline Engine/Login/LoginCliAction.cs:40]: new lines?

File: Engine/Login/RefreshTokenAction.cs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                if (!response.IsSuccessStatusCode)
+                {
+                    var error = response.Content.ReadAsStringAsync().Result;
+                    throw new Exception("Unable to connect: " + response.StatusCode);
+                }
+
+                var responseString = response.Content.ReadAsStringAsync().Result;
+                var json = System.Text.Json.JsonDocument.Parse(responseString);
+                var accessExp = DateTime.Now.AddSeconds(300);
+                var refreshExp = DateTime.Now.AddSeconds(1600);
```
Comment (db-ks) [inline Engine/Login/RefreshTokenAction.cs:69]: 1800?

File: Engine/Login/RefreshTokenAction.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                if(json.RootElement.TryGetProperty("expires_in", out var exp1Str) && exp1Str.TryGetInt32(out var accessAdd))
+                    accessExp = DateTime.Now.AddSeconds(accessAdd - 5);
```
Comment (db-ks) [inline Engine/Login/RefreshTokenAction.cs:71]: What happens if 5 seconds do not suffice due to network issues, service down, etc?

File: Engine/Login/RefreshTokenAction.cs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+                if(json.RootElement.TryGetProperty("refresh_expires_in", out exp1Str) && exp1Str.TryGetInt32(out var refreshAdd))
+                    refreshExp = DateTime.Now.AddSeconds(refreshAdd - 5 );
+                
+                var accessToken = json?.RootElement.GetProperty("access_token").GetString();
+                if(accessToken != null)
+                    LoginInfo.Current.RegisterAccessToken(Site, accessToken, accessExp);
+                refreshToken = json?.RootElement.GetProperty("refresh_token").GetString();
+                if(refreshToken != null)
```
Comment (db-ks) [inline Engine/Login/RefreshTokenAction.cs:79]: This part can be refactored

File: Engine/Login/TokenInfo.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    public class TokenInfo
+    {
+        /// <summary> Raw token string. </summary>
+        public string TokenData { get; set; }
+        /// <summary> Expiration date of the token. </summary>
+        public DateTime Expiration { get; set; }
+        /// <summary> The site this token activates. </summary>
+        public string Site { get; set; }
+        /// <summary> The type of token. </summary>
+        public string Type { get; set; }
```
Comment (db-ks) [inline Engine/Login/TokenInfo.cs:15]: could `enum` instead of `string` be a better choice?

File: Engine/Login/RefreshTokenAction.cs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+                   {
+                       {"refresh_token", refreshToken},
+                       {"grant_type", "refresh_token"}, 
+                       {"client_id", clientId}
+                   }))
+            {
+                var response = http.PostAsync(url, content, cancellationToken).Result;
+                if (!response.IsSuccessStatusCode)
+                {
+                    var error = response.Content.ReadAsStringAsync().Result;
```
Comment (db-ks) [inline Engine/Login/RefreshTokenAction.cs:62]: What is `error` used for?

File: Engine/Login/TokenInfo.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+{
+    /// <summary> Represents stored information about a token. </summary>
+    public class TokenInfo
+    {
+        /// <summary> Raw token string. </summary>
+        public string TokenData { get; set; }
+        /// <summary> Expiration date of the token. </summary>
+        public DateTime Expiration { get; set; }
+        /// <summary> The site this token activates. </summary>
+        public string Site { get; set; }
```
Comment (db-ks) [inline Engine/Login/TokenInfo.cs:13]: What do you mean by site?

File: Package/Repositories/HttpPackageRepository.cs [hunk 9]
```diff
@@ -242,6 +240,9 @@ private string downloadPackagesString(string args, string data = null, string co
                     wc.Proxy = WebRequest.GetSystemWebProxy();
                     wc.Headers.Add(HttpRequestHeader.Accept, accept ?? "application/xml");
                     wc.Headers.Add("OpenTAP", PluginManager.GetOpenTapAssembly().SemanticVersion.ToString());
+                    var token = LoginInfo.Current.GetValidAccessToken(new Uri(Url).Host, TapThread.Current.AbortToken);
+                    if (token != null)
```
Comment (db-ks) [inline Package/Repositories/HttpPackageRepository.cs:244]: You could consider using `HttpClient` (and AuthenticationHeaderValue).

File: Engine/Login/LoginCliAction.cs [hunk 10]
```diff
@@ -0,0 +1,72 @@
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading;
+using OpenTap.Cli;
+
+namespace OpenTap.Login
+{
+    /// <summary>
+    /// Logs in to an Keycloak instance.
```
Comment (db-ks) [inline Engine/Login/LoginCliAction.cs:10]: Is this only meant to be used against Keycloak? If `yes` shouldn't we highlight it in the name?

File: Engine/Login/LoginCliAction.cs [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+        /// <summary> This can be used if the url does not contain the service site. </summary>
+        [CommandLineArgument("site")] public string Site { get; set; }
+
+        private static readonly TraceSource log = Log.CreateSource("web");
+
+        /// <summary> Executes the action. </summary>
+        public int Execute(CancellationToken cancellationToken)
+        {
+            var url = $"https://{Url}/auth/realms/{Realm}/protocol/openid-connect/token";
```
Comment (db-ks) [inline Engine/Login/LoginCliAction.cs:34]: `auth/realms` seems specific for Keycloak.
`https://{Url}/auth/realms/{Realm}/` seems generic however.

File: Engine/Login/LoginCliAction.cs [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+                   }))
+            {
+                var response = http.PostAsync(url, content, cancellationToken).Result;
+                if (!response.IsSuccessStatusCode)
+                {
+                    var error = response.Content.ReadAsStringAsync().Result;
+                    throw new Exception("Unable to connect: " + response.StatusCode);
+                }
+
+                Site = Site ?? Url.Replace("keycloak.", "");
```
Comment (db-ks) [inline Engine/Login/LoginCliAction.cs:50]: Why do you need to do this replace?

File: Engine/Login/LogoutAction.cs [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+        /// <summary> The URL for the keycloak instance</summary>
+        [CommandLineArgument("url")] public string Url { get; set; }
+        
+        private static readonly TraceSource log = Log.CreateSource("web");
+        /// <summary>  Executes the action. </summary>
+        public int Execute(CancellationToken cancellationToken)
+        {
+            if (Url == null)
+                throw new ArgumentException("Not set", nameof(Url));
+            LoginInfo.Current.UnregisterAccessToken(Url.Replace("keycloak.", ""));
```
Comment (db-ks) [inline Engine/Login/LogoutAction.cs:20]: Why do you need to perform this replace?

**Review by db-ks — COMMENTED** [review]
Looks better to me!

**Review by alnlarsen — COMMENTED** [review]
The code looks good, but I'm a bit concerned with the number of public members we are adding. Do they all need to be public?

**Pull request merged**